rc/tools/lint.kak

```diff
@@ -1,113 +1,339 @@
+        The shell command used by lint-buffer and lint-selections.
+
+        It will be given the path to a file containing the text to be
+        linted, and must produce output in the format:
+
+            {filename}:{line}:{column}: {kind}: {message}
+
+        If the 'kind' field contains 'error', the message is treated
+        as an error, otherwise it is assumed to be a warning.
+    } \
+    str lintcmd
 
 declare-option -hidden line-specs lint_flags
-declare-option -hidden range-specs lint_errors
+declare-option -hidden line-specs lint_messages
 declare-option -hidden int lint_error_count
 declare-option -hidden int lint_warning_count
 
-define-command lint -docstring 'Parse the current buffer with a linter' %{
+define-command \
+    -hidden \
+    -params 1 \
+    -docstring %{
+        lint-cleaned-selections <linter>: Check each selection with <linter>.
+
+        Assumes selections all have anchor before cursor, and that
+        %val{selections} and %val{selections_desc} are in the same order.
+    } \
+    lint-cleaned-selections \
+%{
+    # Clear the current contents of the various options.
+    set-option buffer lint_flags %val{timestamp}
+    set-option buffer lint_messages %val{timestamp}
+    set-option buffer lint_error_count 0
+    set-option buffer lint_warning_count 0
+
+    # Create a temporary directory to keep all our state.
     evaluate-commands %sh{
-        if [ -z "${kak_opt_lintcmd}" ]; then
-            echo 'fail The `lintcmd` option is not set'
-            exit 1
-        fi
+        # This is going to come in handy later.
+        kakquote() {
+            printf "'"
+            printf "%s" "$*" | sed "s/'/''/g"
+            printf "'"
+        }
+
+        # Before we clobber our arguments,
+        # let's record the lintcmd we were given.
+        lintcmd="$1"
 
+        # Some linters care about the name or extension
+        # of the file being linted, so we'll store the text we want to lint
+        # in a file with the same name as the original buffer.
         filename="${kak_buffile##*/}"
 
+        # A directory to keep all our temporary data.
         dir=$(mktemp -d "${TMPDIR:-/tmp}"/kak-lint.XXXXXXXX)
-        mkfifo "$dir"/fifo
-        printf '%s\n' "evaluate-commands -no-hooks write -sync $dir/${filename}"
 
+        # A fifo to send the results back to a Kakoune buffer.
+        # FIXME: Should we put the lint output in toolsclient?
+        mkfifo "$dir"/fifo
         printf '%s\n' "evaluate-commands -draft %{
-                  edit! -fifo $dir/fifo -debug *lint-output*
+                  edit! -fifo $(kakquote "$dir/fifo") -debug *lint-output*
                   set-option buffer filetype make
                   set-option buffer make_current_error_line 0
-                  hook -always -once buffer BufCloseFifo .* %{ nop %sh{ rm -r '$dir' } }
               }"
 
+        # Write all the selection descriptions to files.
+        eval set -- "$kak_quoted_selections_desc"
```

Comment:
Technically no, because `selections_desc` should never have any unwanted shell-sensitive characters in it. However, I find it easier to always look for the `_quoted` keyword than to try and remember whether a given expansion happens to be shell-safe or not.

Reply:
Since it takes extra time to produce a quoted value, and we're discussing saving a dozen milliseconds on `kakquote()`, maybe we could remove `_quoted` here?

Reply:
Done.